### PR TITLE
Prophetnet batch dimension inversion fix

### LINKED ISSUE #17455: ProphetNet inconsistent with changing batch ordering
### System Info

```shell
- `transformers` version: 4.15.0
- Platform: Windows-10-10.0.19044-SP0
- Python version: 3.9.7
- PyTorch version (GPU?): 1.10.1+cpu (False)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: No
- Using distributed or parallel set-up in script?: No
```

### Who can help?

@patrickvonplaten

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction

``` python
from transformers import ProphetNetForConditionalGeneration, ProphetNetTokenizer

model = ProphetNetForConditionalGeneration.from_pretrained('microsoft/prophetnet-large-uncased')
tokenizer = ProphetNetTokenizer.from_pretrained('microsoft/prophetnet-large-uncased')

input_string = ['Hello, my dog is cute', 'I have a cat that is not cute']
labels = ['My dog is cute', 'My cat is not cute']

inputs = tokenizer(input_string, return_tensors="pt", padding=True, truncation=True)
targets = tokenizer(labels, return_tensors="pt", padding=True, truncation=True)

# Inverse the ordering of the input and labels using [::-1]
inputs_inv = tokenizer(input_string[::-1], return_tensors="pt", padding=True, truncation=True)
targets_inv = tokenizer(labels[::-1], return_tensors="pt", padding=True, truncation=True)

output = model(input_ids=inputs.input_ids, attention_mask=inputs.attention_mask, labels=targets.input_ids)
output_inv = model(input_ids=inputs_inv.input_ids, attention_mask=inputs_inv.attention_mask, labels=targets_inv.input_ids)

print(output.loss.item())
print(output_inv.loss.item())
```
Given two different forward passes, where one of them is the inverse of the other, the two losses are different.

output.loss.item() = 5.023777484893799
output_inv.loss.item() = 6.5036540031433105

When comparing their encoder output (last hidden states), they are again not equal
```Python
enc_output = model.prophetnet.encoder(input_ids=inputs.input_ids, attention_mask=inputs.attention_mask)
enc_output_inv = model.prophetnet.encoder(input_ids=inputs_inv.input_ids, attention_mask=inputs_inv.attention_mask)
# We flip the inverse encoder output to have the same order
print((enc_output.last_hidden_state == enc_output_inv.last_hidden_state.flip(0)).all())
``` 
Also equals `False`

### Expected behavior

```shell
I expect the order of the batch to not have an influence on the model output (aside from the ordering of the output)
```

**Edit** The issue persistent regardless of padding and also causes model.generate to generate two different sets of text.

### PR DESCRIPTION
**Pull request by kiansierra**
# What does this PR do?

Fixes #17455

No longer mixes batch dimension with other dimensions, makes it so that inverting the inputs along the batch dimension also inverts the outputs, and returns the same loss indepent of the batch order.
Currently all tests pass (locally) except integration tests, which I believe are due to the issue at hand as can be seen in this example. Essentially the integration test for generation, returns different outputs based on what other elemnts are in the batch, with this fix it returns the same output as with a batch of 1 
[Open In Colab](https://colab.research.google.com/drive/12EAAbXZSemzvuoz5g_3WAe0sH1YAZUwk?usp=sharing)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [X] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. 
@patrickvonplaten @patil-suraj

### CODE REVIEW

**Comment by sgugger** [comment]
cc @younesbelkada and @ArthurZucker

File: tests/models/convbert/test_modeling_convbert.py [hunk 1]
```diff
@@ -459,6 +459,11 @@ def test_model_for_input_embeds(self):
         result = model(inputs_embeds=inputs_embeds)
         self.assertEqual(result.last_hidden_state.shape, (batch_size, seq_length, config.hidden_size))
 
+    def test_reducing_attention_heads(self):
```
Comment (ArthurZucker) [inline tests/models/convbert/test_modeling_convbert.py:462]: this should not be included in the PR, please revert all changes unrelated

**Review by ArthurZucker — COMMENTED** [review]
Thanks for working on this! Let's try to make the addition better for the code overall 😉 by that I mean let's use comments that are descriptive and let's try to split in two lines where we can instead of having some `view` spread.

File: tests/models/convbert/test_modeling_convbert.py (hunk 1 above)
Comment (kiansierra) [reply tests/models/convbert/test_modeling_convbert.py]: It's been removed, this got mixed in when trying to rebase

File: src/transformers/models/convbert/modeling_convbert.py [hunk 2]
```diff
@@ -316,7 +316,7 @@ def __init__(self, config):
         if config.hidden_size % self.num_attention_heads != 0:
             raise ValueError("hidden_size should be divisible by num_attention_heads")
 
-        self.attention_head_size = config.hidden_size // config.num_attention_heads
```
[the remark this answers is not in the document]
Comment (kiansierra) [reply src/transformers/models/convbert/modeling_convbert.py]: It's been removed, this got mixed in when trying to rebase

File: src/transformers/generation/flax_utils.py [hunk 3]
```diff
@@ -448,10 +448,11 @@ def generate(
...
 
-            if "attention_mask" in model_kwargs:
-                model_kwargs["attention_mask"] = self._expand_to_num_beams(
-                    model_kwargs["attention_mask"], num_beams=generation_config.num_beams
-                )
+            for kwarg in ["attention_mask", "decoder_attention_mask"]:
+                if kwarg in model_kwargs:
+                    model_kwargs[kwarg] = self._expand_to_num_beams(
+                        model_kwargs[kwarg], num_beams=generation_config.num_beams
+                    )
```
[the remark this answers is not in the document]
Comment (kiansierra) [reply src/transformers/generation/flax_utils.py]: It's been removed, this got mixed in when trying to rebase

File: src/transformers/models/prophetnet/modeling_prophetnet.py [hunk 4]
```diff
@@ -997,37 +986,35 @@ def forward(
...
         )
         # project to attention output
-        # [ngram, B*head, T, c]
-        predict_attn_output = torch.einsum("nbts,nbsc->nbtc", (predict_attn_probs, predict_value_states))
+        # [b n nh t x] x [b n nh x  hd]-> [b n nh t hd]
+        # return predict_attn_probs, predict_value_states
+        predict_attn_output = torch.einsum(
+            "bnhts,bnhsc->bnhtc", (predict_attn_probs, predict_value_states.transpose(1, 2))
+        )
+        # return predict_attn_probs, predict_value_states.transpose(1,2), predict_attn_output
```
[the remark this answers is not in the document]
Comment (kiansierra) [reply src/transformers/models/prophetnet/modeling_prophetnet.py]: removed (were part of ensuring forward pass is consistent)

File: src/transformers/models/prophetnet/modeling_prophetnet.py [hunk 5]
```diff
@@ -997,37 +986,35 @@ def forward(
...
+        # [B, (1+ngram)*T, hs]
         attn_output = torch.cat([main_attn_output, predict_attn_output], 1).view(batch_size, -1, hidden_size)
         # reshape into better form for `config.output_attentions`
         main_attn_probs = main_attn_probs.view(batch_size, self.num_attn_heads, sequence_length, -1)
-        predict_attn_probs = predict_attn_probs.view(
-            self.ngram, batch_size, self.num_attn_heads, sequence_length, -1
-        ).transpose(0, 1)
+        # predict_attn_probs = predict_attn_probs.view(
+        #     self.ngram, batch_size, self.num_attn_heads, sequence_length, -1
+        # ).transpose(0, 1)
```
[the remark this answers is not in the document]
Comment (kiansierra) [reply src/transformers/models/prophetnet/modeling_prophetnet.py]: removed

File: src/transformers/models/prophetnet/modeling_prophetnet.py [hunk 6]
```diff
@@ -997,37 +986,35 @@ def forward(
...
-            )
+            predict_attn_probs = layer_head_mask.view(1, 1, -1, 1, 1) * predict_attn_probs
 
         predict_attn_probs = nn.functional.dropout(
             predict_attn_probs, p=self.attention_dropout, training=self.training
         )
         # project to attention output
-        # [ngram, B*head, T, c]
-        predict_attn_output = torch.einsum("nbts,nbsc->nbtc", (predict_attn_probs, predict_value_states))
+        # [b n nh t x] x [b n nh x  hd]-> [b n nh t hd]
```
[the remark this answers is not in the document]
Comment (kiansierra) [reply src/transformers/models/prophetnet/modeling_prophetnet.py]: updated to full text notation

File: src/transformers/models/prophetnet/modeling_prophetnet.py [hunk 7]
```diff
@@ -1055,12 +1043,13 @@ def get_main_relative_pos_embeddings(
...
 
         rel_pos_embeddings = self.relative_pos_embeddings(hidden_states)  # [B,T,Buckets*head]
+        # return rel_pos_embeddings
         rel_pos_embeddings = rel_pos_embeddings.view(
             rel_pos_embeddings.shape[:2] + (self.num_buckets, self.num_attn_heads)
         ).permute(
             0, 3, 1, 2
         )  # [B,T,Buckets,head]
-        rel_pos_embeddings = rel_pos_embeddings.reshape(attn_weights.shape[:2] + (-1,))  # [B*head,T,Buckets]
+        rel_pos_embeddings = rel_pos_embeddings.reshape(attn_weights.shape[:3] + (-1,))  # [B, head,T,Buckets]
```
[the remark this answers is not in the document]
Comment (kiansierra) [reply src/transformers/models/prophetnet/modeling_prophetnet.py]: updated to full text notation

File: src/transformers/models/prophetnet/modeling_prophetnet.py [hunk 8]
```diff
@@ -1100,26 +1094,25 @@ def get_predict_relative_pos_embeddings(
...
-        rel_pos_embeddings = rel_pos_embeddings.reshape(-1, rel_pos_embeddings.size(-1))
         predict_relative_position_buckets = predict_relative_position_buckets.view(
             -1, predict_relative_position_buckets.size(-1)
         ).long()  # [ngram*B*head*T, S]
 
         predict_relative_pos_embeddings = torch.gather(
             rel_pos_embeddings, dim=1, index=predict_relative_position_buckets
-        ).view(
-            self.ngram, batch_size * self.num_attn_heads, sequence_length, -1
+        )
```
[the remark this answers is not in the document]
Comment (kiansierra) [reply src/transformers/models/prophetnet/modeling_prophetnet.py]: separated gather and view operation each in one line

File: src/transformers/models/prophetnet/modeling_prophetnet.py [hunk 9]
```diff
@@ -1100,26 +1094,25 @@ def get_predict_relative_pos_embeddings(
...
         )  # [ngram, B, T, bucket, head]
-        rel_pos_embeddings = rel_pos_embeddings.permute(0, 1, 4, 2, 3).reshape(
-            self.ngram * batch_size * self.num_attn_heads, sequence_length, -1
-        )  # [ngram*B*head, T, bucket]
-
+        rel_pos_embeddings = rel_pos_embeddings.permute(0, 2, 1, 4, 3)
+        rel_pos_embeddings = rel_pos_embeddings.reshape(-1, self.num_buckets)  # [ngram*B*head, T, bucket]
         predict_relative_position_buckets = predict_relative_position_buckets.unsqueeze(0).repeat(
             self.ngram, 1, self.num_attn_heads, 1
         )  # [ngram, B, head*T, S]
```
[the remark this answers is not in the document]
Comment (kiansierra) [reply src/transformers/models/prophetnet/modeling_prophetnet.py]: separated permute and reshape operations

File: src/transformers/models/prophetnet/modeling_prophetnet.py [hunk 10]
```diff
@@ -1100,26 +1094,25 @@ def get_predict_relative_pos_embeddings(
...
         ).long()  # [ngram*B*head*T, S]
 
         predict_relative_pos_embeddings = torch.gather(
             rel_pos_embeddings, dim=1, index=predict_relative_position_buckets
-        ).view(
-            self.ngram, batch_size * self.num_attn_heads, sequence_length, -1
+        )
+        predict_relative_pos_embeddings = predict_relative_pos_embeddings.view(
+            batch_size, self.ngram, self.num_attn_heads, sequence_length, -1
         )  # [ngram, B*head, T, S]
```
[the remark this answers is not in the document]
Comment (kiansierra) [reply src/transformers/models/prophetnet/modeling_prophetnet.py]: updated notation to full text and put above line

File: src/transformers/models/prophetnet/modeling_prophetnet.py [hunk 11]
```diff
@@ -1071,15 +1060,20 @@ def get_main_relative_pos_embeddings(
 
         main_relative_pos_embeddings = torch.gather(
             rel_pos_embeddings, dim=1, index=main_relative_position_buckets
-        ).view(attn_weights.shape[:2] + (-1,))
+        ).view(
+            batch_size,
+            num_attn_heads,
+            tgt_len,
+            -1,
+        )
```
[the remark this answers is not in the document]
Comment (kiansierra) [reply src/transformers/models/prophetnet/modeling_prophetnet.py]: split in 2

**Comment by kiansierra** [comment]
Hi @ArthurZucker, thanks for the feedback.
I've implemented the suggestions you mentioned, adding full text notation of expected tensor dimensions and separating tenosr operations in two multiple lines instead of chaining where requested

File: src/transformers/models/prophetnet/modeling_prophetnet.py [hunk 12]
```diff
@@ -721,24 +721,19 @@ def forward(
...
         if attention_mask is not None:  # don't attend to padding symbols
             attn_weights = attn_weights + attention_mask
-
         if output_attentions:
-            # this operation is a bit awkward, but it's required to
-            # make sure that attn_weights keeps its gradient.
-            # In order to do so, attn_weights have to be reshaped
-            # twice and have to be reused in the following
-            attn_weights_reshaped = attn_weights.view(batch_size, self.num_attn_heads, tgt_len, src_len)
-            attn_weights = attn_weights_reshaped.view(batch_size * self.num_attn_heads, tgt_len, src_len)
```
Comment (ArthurZucker) [inline src/transformers/models/prophetnet/modeling_prophetnet.py:741]: love to this this go away! 😉

File: src/transformers/models/prophetnet/modeling_prophetnet.py [hunk 13]
```diff
@@ -701,15 +701,15 @@ def forward(
...
+        proj_shape = (batch_size, self.num_attn_heads, -1, self.head_dim)
         query_states = self._shape(query_states, tgt_len, batch_size).view(*proj_shape)
         key_states = key_states.view(*proj_shape)
         value_states = value_states.view(*proj_shape)
-
-        src_len = key_states.size(1)
-        attn_weights = torch.bmm(query_states, key_states.transpose(1, 2))
+        src_len = key_states.size(2)
+        attn_weights = torch.einsum("bsij,bsjk->bsik", query_states, key_states.transpose(2, 3))
         assert attn_weights.size() == (
```
Comment (ArthurZucker) [inline src/transformers/models/prophetnet/modeling_prophetnet.py:710]: It's a small nit but let's raise an error here instead since we are touching this!

File: src/transformers/models/prophetnet/modeling_prophetnet.py [hunk 14]
```diff
@@ -721,24 +721,19 @@ def forward(
         if attention_mask is not None and attention_mask.dim() == 0:
             attention_mask = None
         assert attention_mask is None or attention_mask.size() == (
-            self.num_attn_heads * batch_size,
```
Comment (ArthurZucker) [inline src/transformers/models/prophetnet/modeling_prophetnet.py:724]: same comment here, will make it cleaner!

File: src/transformers/models/prophetnet/modeling_prophetnet.py [hunk 15]
```diff
@@ -762,22 +756,17 @@ def forward(
             p=self.attention_dropout,
             training=self.training,
         )
-
-        attn_output = torch.bmm(attn_probs, value_states)
+        attn_output = torch.einsum("bsij,bsjk->bsik", attn_probs, value_states)
         assert attn_output.size() == (
```
Comment (ArthurZucker) [inline src/transformers/models/prophetnet/modeling_prophetnet.py:760]: same here, let's raise an error

File: src/transformers/models/prophetnet/modeling_prophetnet.py [hunk 16]
```diff
@@ -895,28 +881,29 @@ def forward(
...
         # saved states are stored with shape (batch_size, num_attn_heads, seq_len, head_dim)
         if past_key_value is not None:
-            prev_main_key_states = past_key_value[0].view(batch_size * self.num_attn_heads, -1, self.head_dim)
-            main_key_states = torch.cat((prev_main_key_states, main_key_states), dim=1)
-            prev_main_value_states = past_key_value[1].view(batch_size * self.num_attn_heads, -1, self.head_dim)
-            main_value_states = torch.cat((prev_main_value_states, main_value_states), dim=1)
+            prev_main_key_states = past_key_value[0]
+            main_key_states = torch.cat((prev_main_key_states, main_key_states), dim=2)
+            prev_main_value_states = past_key_value[1]
+            main_value_states = torch.cat((prev_main_value_states, main_value_states), dim=2)
```
Comment (ArthurZucker) [inline src/transformers/models/prophetnet/modeling_prophetnet.py:887]: nice

**Review by ArthurZucker — APPROVED** [review]
Okay! This looks a lot better thanks for your contribution. Pinging @sgugger for another eye on this

**Comment by kiansierra** [comment]
Thanks for the kind feedback @ArthurZucker.
Just some clarity before merging, I should update the integration tests as described by the attached colab, the current version generates different text based on the other elements in the batch. While new version returns the same output as if generated with a batch size of 1

**Comment by kiansierra** [comment]
I've now updated the integration tests, they should pass now

**Review by sgugger — APPROVED** [review]
Nice, thanks for the fix!
